doc/HLD_Data_Block_Allocator.rst

```diff
@@ -0,0 +1,407 @@
+The data block allocator manages the free spaces in the container, and provides "allocate" and "free" blocks interfaces to other components and layers. The main purpose of this document is to provide an efficient algorithm to allocate and free blocks quickly, transactionally and friendly to I/O performance. The mapping from logical block number in sub-components to physical block number is out of the scope of this document. That is another task, and will be discussed in another document.
+
+******************
+Definitions
+******************
+
+- Extent. Extent is used to describe a range of space, with "start" block number and "count" of blocks.
+
+- Block. The smallest unit of allocation.
+
+- Block Size. The number of bytes of a Block.
+
+- Group. The whole space is equally divided into groups with fixed size. That means every group has the same amount of blocks. When allocating spaces, groups are iterated to search for the best candidate. Group is locked during this step. Having multiple groups in a container can reduce the lock contention.
+
+- Super Group. Group 0 is an special group. It contains metadata of this container, and is not allocatable to ordinary objects.
+
+*************
+Requirements
+*************
+
+- The data-block-allocator should perform well for large continuous I/O, small object I/O.
+
+- The data-block-allocator should survive node crash.
+
+- DB5 should be used to track the block usage.
+
+- The allocator should have similar strategy as the ext4 allocator.
+
+- The allocator should be designed for concurrency by multiple processes
+
+- The allocator should support container inclusion
+
+- The allocator should support merging allocation data of sub-containers into that of their parents
+
+- The allocator should leave FOL traces sufficient to support FOL driven fsck plugins which support all important operations fsck normally provides.
+
+- Pre-allocation is supported
+
+*****************
+Design Highlights
+*****************
+
+C2 data-block-allocator will use the same algorithm as that from ext4 to do block allocation. But instead of using bitmap to track the block usage, C2 will use extent to track the free space. These free space extent will be stored in database, and updated with transaction support. Highlights in the C2 allocator (or derived from ext4 allocator) are:
+
+- Container is equally divided into groups, which can be locked respectively.
+
+- Extents are used to track the free spaces in groups.
+
+- Every group has a statistic meta-data about this group, such as largest available extent, total free block count.
+
+- All these information are stored in databases. We will use Oracle DB5 in our project.
+
+- Object-based pre-allocation and group-based pre-allocation are both supported.
+
+- Blocks are allocated in a bunch. Multi-blocks can be allocated at one time. Buddy algorithm is used to do allocation.
+
+***************************
+Functional Specification
+***************************
+
+C2_mb_format
```

Comment:
Relace the "C2_" prefix with "m0_" prefix.

Reply:
Incorporated.